### ISSUE TITLE
Training and evaluating on new dataset

### ISSUE THREAD

**Issue by lelouedec** [ShivamDuggal4/TARS3D#2]
Good morning, 
Congrats on the very interesting paper! And very glad to read the nice code after seeing your CVPR poster.
I have one dataset composed of 3D meshes and their render. And would like to train and evaluate your method for comparison with our work. 
If I understood well I could copy one of the scripts folder and adapt it ? 
But should I rewrite all the dataloader classes and script or is there a fastest and simple way to do it. 
The dataset is quite simple : mesh + renders + pose and camera parameters.

Thank you in advance :)
Justin

**Comment by ShivamDuggal4** [comment]
Hi @lelouedec 

Thanks for your interest in our work and thanks for the warm words!

You can check any of the dataloader files and try to adapt it to the new dataset, maybe check the [Shapenet dataloader](https://github.com/ShivamDuggal4/TARS3D/blob/master/dataloader/shapenet.py).

Few points to ensure while creating the dataloader would be:

- Your dataloader file will require object images, masks and camera poses and will return a dictionary of {images, masks, camera poses, intrinsics, distance_transform, point cloud (for validation only)}.
- We learn the object shape within a bounded cube. The bounds of that cube (generally [-1,1]) are provided via a config parameter (eg: as mentioned [here](https://github.com/ShivamDuggal4/TARS3D/blob/6988d4e24442f8e0891166ba3516cdb2d7ee1582/configs/shapenet/base.yaml#L43)). To ensure that the network learns shapes within this bound, the following things need to be ensured: (a) your renderings (images and masks) need to be tightly cropped around the object of interest [(see this for reference)](https://github.com/ShivamDuggal4/TARS3D/blob/6988d4e24442f8e0891166ba3516cdb2d7ee1582/dataloader/pascal3d.py#L107), (b) point cloud for validation need to be scaled to [-1,1] [(see this for reference)](https://github.com/ShivamDuggal4/TARS3D/blob/6988d4e24442f8e0891166ba3516cdb2d7ee1582/dataloader/pascal3d.py#L155) for reference).
- I would recommend setting up the dataloader (keeping in mind the points mentioned above) and then testing that everything is as expected by shooting rays from GT camera viewpoints (using functions defined [here](https://github.com/ShivamDuggal4/TARS3D/blob/6988d4e24442f8e0891166ba3516cdb2d7ee1582/misc/camera.py#L112)) and seeing if they interest the GT mesh correctly.

Good luck with your testing!

Best Regards
Shivam

**Comment by lelouedec** [comment]
Hi @ShivamDuggal4 

I followed your advice and created a data loader to provide what was necessary. I the of the functions you provide to obtain most of this information, and rotated the point clouds to match the image view point. 
The images should correspond to the description and the point clouds/meshs are normalized between -1 and 1. 
Do you have examples of the evaluation visdom visualization which should be expected for reconstruction stage one? 
![image](https://user-images.githubusercontent.com/16105109/187209069-5d0677f0-cff5-415a-a9f1-67c124864fc0.png)
This is what I am getting, the blue prediction is rather flat. 
Thank you again for the help :)

**Comment by weiyao1996** [comment]
Hi all.
@ShivamDuggal4 thanks for your codes of this great work. Also, thank you @lelouedec  for opening this issue. Actually, I have a similar question, i.e., I would like to train and evaluate on a new dataset. Different from the dataset of @lelouedec, in my project, I only have object images, masks and 3D point clouds. I am not sure if this method can be applied on my dataset since there is no pose information. To fit this method, could you please give me some suggestions on generating camera pose for each image? If I understand correctly, it should be more appropriate for me to develop based on the pascal3d dataloader, right?
Many thanks in advance.

Best,
Yao

**Comment by lelouedec** [comment]
hey , 

I think I might have made a mistake now that I am considering the pose similarly to your problem. 
I am not 100%, but the camera pose is important, not the object pose. So in my case, I shouldn't rotate my objects, but rather the camera. So I fixed that and here are the inputs: 
![2022-08-30-092115_1667x488_scrot](https://user-images.githubusercontent.com/16105109/187375033-aa90a255-a6d9-412e-b864-dcc91186be91.png)
All the input seems correct (also tried to ray trace all pixels and it works as expected) and the camera pose and intrisics seem fine (example with no rotation, ~2.0 on Z and facing 0,0,0, and focal of 1.0472), I tried rendering a mesh with pyrender and it was correct : 
'pose': tensor([[-1.          0.          0.         -0.        ]
                       [-0.         -0.99986128 -0.0166562  -0.03331239]
                       [ 0.         -0.0166562   0.99986128  1.99972255]
                       [ 0.          0.          0.          1.        ]]
                      ), 
'intr': tensor([[67.0208,  0.0000, 32.0000],
        [ 0.0000, 67.0208, 32.0000],
        [ 0.0000,  0.0000,  1.0000]])

But stage one, just keep producing weird shapes like that : 
![image](https://user-images.githubusercontent.com/16105109/187386312-e7a5e719-74e6-45ba-8193-b25d0e62c34b.png)

For @weiyao1996 and pose, I would say if you have a canonical pose for the object (like a vertical neutral pose), you can use as a reference and compute the camera angle for it, this is the best way. For my example I use the vertical strawberry pose as reference and compute all cameras from that.

EDIT: I realized reading the code of visualization that there was a tensorboard with more images. And it seems that depth and level just don't learn much :/ : 
![image](https://user-images.githubusercontent.com/16105109/187998352-915e538f-a7c7-44fd-a146-006dc9df4665.png)

**Comment by ShivamDuggal4** [comment]
Hi @lelouedec 

From my experiments with several different object categories, stage 1 never produced flat shape if the camera poses are correct. To debug the issue, maybe try the following:
- To confirm, are you running the pre-training step before stage 1 ?
- I would recommend comparing the ShapeNet mesh/ camera poses with your dataset mesh/ camera poses, and try to bring them in the same coordinate system. That's how I setup all the other datasets after I had setup the ShapeNet one.
- Your dataset seems like a symmetric shape. If the shape you are trying to reconstruct is like a completely symmetric shape with largely homogeneous texture (such that the object image rendered from all viewpoints look exactly the same), then it might be difficult for the network to learn a meaningful 3D shape.

Best Regards
Shivam

**Comment by ShivamDuggal4** [comment]
Hi @weiyao1996 

- Since your dataset has object images, masks, and point clouds, you can try generating the camera poses by registering (projecting/ rendering) the point clouds with the captured images/ masks. If your point clouds are textured, you might be able to register the point clouds to the images decently.

- The camera poses defined in this work are the transformation of the 3D meshes from the canonical space to the camera coordinates. You can define any pose to be the canonical pose for the entire category, and then define all the camera poses to be relative to that.

- An alternative way to generate camera poses could be: Annotating key points on all captured images and then running an SfM (structure from motion) pipeline on the annotated key points to get rough camera pose estimates. We use the camera poses generated by such a mechanism for the CUBS category (as provided by CMR). Please refer to prior works ([[1]](https://arxiv.org/pdf/1411.6069.pdf), [[2]](https://arxiv.org/pdf/1803.07549.pdf)) for more details.

Best Regards,
Shivam

**Comment by lelouedec** [comment]
Hi @ShivamDuggal4 ,

Thank you for the advice. I spent a few days playing with parameters, using pieces of your code for the dataloader and checking the camera poses. The error came from the camera coordinate system being different from what I was visualizing with pyrenderer. 
I was doing all pretraining and generation steps as you explained in readme, and now I get some results making more sense. 
Concerning the objects, they are strawberries Ahah, and indeed nearly symmetrical. The main problem as well is that every sample is unique compared to birds or chair of the same family/species/type.
I think the first reconstruction stage converged toward something good and now I am working on tweaking config files for the second reconstruction. 
But is it normal that the center of level/distance image is empty ? 
![image](https://user-images.githubusercontent.com/16105109/188472555-e8a5e219-69e8-46df-b0a7-f7699ac335cc.png)

**Comment by ShivamDuggal4** [comment]
Hi @lelouedec 

Sorry for the late reply!
Yes, that is completely fine, given that we use GT distance transform to only optimize the lower bound of the SDF of the **non-surface points** (points which lie outside the 3D surface and hence are projected outside the object's image silhouette).

Best Regards,
Shivam

**Comment by ShivamDuggal4** [comment]
Closing the issue because of inactivity! Hope you managed to get it working on your dataset.
